### PR TITLE
PLAT-1336 Improved ExceptionsCollector Internal Structure and Test

### PR DESCRIPTION
**Pull request by oliver-richers**

### CODE REVIEW

File: platform-common/src/test/java/com/softicar/platform/common/core/exception/ExceptionsCollectorTest.java [hunk 1]
```diff
@@ -1,6 +1,8 @@
 package com.softicar.platform.common.core.exception;
 
+import com.softicar.platform.common.core.logging.Log;
 import com.softicar.platform.common.testing.AbstractTest;
+import java.util.List;
 import org.junit.Test;
 
 public class ExceptionsCollectorTest extends AbstractTest {
```
Comment (alexander-schmidt) [inline platform-common/src/test/java/com/softicar/platform/common/core/exception/ExceptionsCollectorTest.java:8]: This test focuses on `Exception` while the `ExceptionsCollector` API uses `Throwable`. That's odd.

File: platform-common/src/test/java/com/softicar/platform/common/core/exception/ExceptionsCollectorTest.java [hunk 2]
```diff
@@ -30,4 +34,25 @@ public void testWithEmptyCollector() {
...
 		assertEquals("", collector.getMessage());
 	}
+
+	private void generateExceptionsWithSameStacktrace(int count) {
+
+		for (int i = 0; i < count; i++) {
+			try {
+				provokeException("exception #" + i);
+			} catch (Exception exception) {
+				Log.finfo(List.of(exception.getStackTrace()));
```
Comment (alexander-schmidt) [inline platform-common/src/test/java/com/softicar/platform/common/core/exception/ExceptionsCollectorTest.java:44]: I think this line should be removed.

**Review by alexander-schmidt — CHANGES_REQUESTED** [review]
needs update, see comments

File: platform-common/src/main/java/com/softicar/platform/common/core/exception/ExceptionsCollector.java [hunk 3]
```diff
@@ -98,4 +105,31 @@ public void consumeIfNotEmpty(Consumer<String> messageConsumer) {
 			messageConsumer.accept(getMessage());
 		}
 	}
+
+	public static class ExceptionCollection {
```
Comment (alexander-schmidt) [inline platform-common/src/main/java/com/softicar/platform/common/core/exception/ExceptionsCollector.java:109]: This `public static class` could be extracted to a separate file.

[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-common/src/main/java/com/softicar/platform/common/core/exception/ExceptionsCollector.java:109]: Yes, ppl never talk about `Throwable` but always `Exception`...so IMO, name should not be changed.

File: platform-common/src/test/java/com/softicar/platform/common/core/exception/ExceptionsCollectorTest.java (hunk 2 above)
Comment (oliver-richers) [reply platform-common/src/test/java/com/softicar/platform/common/core/exception/ExceptionsCollectorTest.java:44]: right, fixed

File: platform-common/src/main/java/com/softicar/platform/common/core/exception/ExceptionsCollector.java [hunk 4]
```diff
@@ -98,4 +105,31 @@ public void consumeIfNotEmpty(Consumer<String> messageConsumer) {
...
+			this.stackTraceElements = stackTraceElements;
+			this.throwables = new ArrayList<>();
+		}
+
+		public List<StackTraceElement> getStackTraceElements() {
+
+			return Collections.unmodifiableList(stackTraceElements);
+		}
+
+		public List<Throwable> getThrowables() {
```
Comment (alexander-schmidt) [inline platform-common/src/main/java/com/softicar/platform/common/core/exception/ExceptionsCollector.java:125]: Calling `getThrowables()` on an `ExceptionCollection` instance seems particularly odd to me. Would bother me less if the method was `public List<Throwable> getExceptions()` or if the class would have `Throwable` in its name.

File: platform-common/src/test/java/com/softicar/platform/common/core/exception/ExceptionsCollectorTest.java (hunk 1 above)
Comment (oliver-richers) [reply platform-common/src/test/java/com/softicar/platform/common/core/exception/ExceptionsCollectorTest.java:8]: so what to do? I don't want to talk about `Throwable` when most people only know about `Exception`

File: platform-common/src/main/java/com/softicar/platform/common/core/exception/ExceptionsCollector.java (hunk 3 above)
Comment (oliver-richers) [reply platform-common/src/main/java/com/softicar/platform/common/core/exception/ExceptionsCollector.java]: ok, I completely redesigned the code an removed the collection class ;)

File: platform-common/src/test/java/com/softicar/platform/common/core/exception/ExceptionsCollectorTest.java (hunk 1 above)
Comment (alexander-schmidt) [reply platform-common/src/test/java/com/softicar/platform/common/core/exception/ExceptionsCollectorTest.java:8]: Well I guess we can't do much as long as we uphold the assumption about ppl being ignorant towards the existence of `Throwable`. Feel free to close this.

File: platform-common/src/main/java/com/softicar/platform/common/core/exception/ExceptionsCollector.java (hunk 3 above)
Comment (alexander-schmidt) [reply platform-common/src/main/java/com/softicar/platform/common/core/exception/ExceptionsCollector.java]: nice

File: platform-common/src/main/java/com/softicar/platform/common/core/exception/MultiExceptionMessageBuilder.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+	public MultiExceptionMessageBuilder(Collection<Throwable> exceptions) {
+
+		this.exceptions = exceptions;
+	}
+
+	public String buildMessage() {
+
+		StringWriter buffer = new StringWriter();
+		try (PrintWriter printer = new PrintWriter(buffer)) {
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-common/src/main/java/com/softicar/platform/common/core/exception/MultiExceptionMessageBuilder.java:22]: ok

File: platform-common/src/main/java/com/softicar/platform/common/core/exception/MultiExceptionMessageBuilder.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+			for (var exceptionList: getExceptionLists()) {
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-common/src/main/java/com/softicar/platform/common/core/exception/MultiExceptionMessageBuilder.java:23]: did this similar

File: platform-common/src/main/java/com/softicar/platform/common/core/exception/MultiExceptionMessageBuilder.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+				printer.append("\n\n");
+				printer.printf("Gathered %s exceptions with the same stacktrace.", exceptionList.size());
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-common/src/main/java/com/softicar/platform/common/core/exception/MultiExceptionMessageBuilder.java:25]: added a separator like `-------------------- 3 exceptions with stack trace #0 --------------------`

**Review by alexander-schmidt — APPROVED** [review]

**Pull request merged**